File: kernel/ps2.c
#include <kernel/event.h>
#include <kernel/io.h>
#include <kernel/isr.h>
#include <kernel/ps2.h>
#include <kernel/trapframe.h>

#include <stdint.h>
#include <stdio.h>

void init_ps2_keyboard() {
    install_interrupt_handler(1, ps2_keyboard_interrupt_handler);
}

void ps2_keyboard_interrupt_handler(uint8_t irq, trap_frame_t *trap_frame, uint32_t error) {
    (void) irq;
    (void) trap_frame;
    (void) error;
    uint8_t status = inb(0x64);
    if (!(status & 1))
        return;
    uint8_t scancode = inb(0x60);
    if (!(scancode & 0x80)) {
        event_t event;
        event.type = KEY_DOWN;
        event.arg0 = scancode;
        new_event(&event);
    } else {
        event_t event;
        event.type = KEY_UP;
        event.arg0 = scancode & 0x7F;
        new_event(&event);
    }
}

File: kernel/src/ps2.zig
const std = @import("std");
const event = @import("event.zig");
const io = @import("io.zig");
const isr = @import("isr.zig");
const kbd = @import("kbd.zig");
const writer = @import("serial.zig").writer;

pub fn initialize() void {
    isr.install_handler(1, interrupt_handler);
}

pub fn interrupt_handler() void {
    if ((io.inb(0x64) & 1) == 0) return;

    const scancode = io.inb(0x60);
    kbd.keyboard_event(scancode);
    var event_params = std.mem.zeroes([8]u32);
    event_params[0] = scancode;
    const key_event = event.Event{
        .event_type = if (scancode & 0x80 != 0) event.EventType.key_up else event.EventType.key_down,
        .parameters = event_params,
    };
    event.new_event(key_event) catch writer.print("failed to create new keyboard event!\n", .{}) catch unreachable;
}
